### PR TITLE
SONARCSANA-130 IDisposable local variables should not be returned fro…

### PR DESCRIPTION
**Pull request by tvsonar**
…m a "using" statement

### CODE REVIEW

File: src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(
+                c =>
+                {
+                    var usingStatement = (UsingStatementSyntax) c.Node;
+                    var declaration = usingStatement.Declaration;
+                    var declaredSymbols = ImmutableArray<ISymbol>.Empty;
+                    if (declaration != null)
```
Comment (dbolkensteyn) [inline src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs]: @tvsonar Does Roslyn use null if the node is not present?

File: src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(Rule, usingStatement.UsingKeyword.GetLocation(),
+                            string.Join(", ",
+                                returnedSymbols.Select(s => string.Format("\"{0}\"", s.Name)))));
+                    }
+                },
+                SyntaxKind.UsingStatement);
+        }
+
+        private static ImmutableArray<ISymbol> GetReturnedSymbolSubset(StatementSyntax usingStatement,
```
Comment (dbolkensteyn) [inline src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs]: @tvsonar This seems quite complex. Why can't you simply return all returned symbols by `return` statements within the `using` block, and then check for the non-empty intersection with `declaredSymbols` ?

(so this function would just take the using statement and the sematic model as parameters)

File: src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs (hunk 1 above)
Comment (tvsonar) [inline src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs]: @dbolkensteyn Yes, either the `Declaration` or the `Expression` is `null`.

File: src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                                returnedSymbols.Select(s => string.Format("\"{0}\"", s.Name)))));
+                    }
+                },
+                SyntaxKind.UsingStatement);
+        }
+
+        private static ImmutableHashSet<ISymbol> GetReturnedSymbols(StatementSyntax usingStatement, 
+            SemanticModel semanticModel)
+        {
+            var enclosingSymbol = semanticModel.GetEnclosingSymbol(usingStatement.SpanStart);
```
Comment (dbolkensteyn) [inline src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs]: @tvsonar If I remove this line (and one more hereunder), all tests are still green - so I think they should be removed

File: src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+            return usingStatement.DescendantNodesAndSelf()
+                .OfType<ReturnStatementSyntax>()
+                .Where(ret => semanticModel.GetEnclosingSymbol(ret.SpanStart).Equals(enclosingSymbol))
```
Comment (dbolkensteyn) [inline src/CSharp.CodeAnalysis/Rules/DisposableReturnedFromUsing.cs]: @tvsonar and this one

**Comment by dbolkensteyn** [comment]
LGTM except for the 2 useless lines

**Comment by tvsonar** [comment]
@dbolkensteyn I added an additional test case that requires these 2 lines

**Pull request merged**